scipy/optimize/tests/test_optimize.py

```diff
@@ -479,6 +527,97 @@ def test_neldermead_adaptive():
+    # One cool feature of the bounded Powell method is that if you
+    # start outside the bounds, the final answer will still be
+    # within the bounds (provided that the user doesn't make a bad
+    # choice for the `direc` argument). 
+    func = lambda x: np.sum(x**2)
+    bounds = (-1, 1), (-1, 1), (-1, 1)
+    x0 = [-4, .5, -.8]
+
+    # we're starting outside the bounds, so we should get a warning
+    with assert_warns(optimize.OptimizeWarning):
+        res = optimize.minimize(func, x0, bounds=bounds, method="Powell")
+    assert_allclose(res.x, np.array([0.] * len(x0)), atol=1e-6)
+    assert_equal(res.success, True)
+    assert_equal(res.status, 0)
+
+    # However, now if we change the `direc` argument such that the
+    # set of vectors does not span the parameter space, then we may
+    # not end up back within the bounds. Here we see that the first
+    # parameter cannot be updated!
+    direc = [[0, 0, 0], [0, 1, 0], [0, 0, 1]]
+    # we're starting outside the bounds, so we should get a warning
+    with assert_warns(optimize.OptimizeWarning):
+        res = optimize.minimize(func, x0, 
+                                bounds=bounds, method="Powell",
+                                options=dict(direc=direc))
+    assert_allclose(res.x, np.array([-4., 0, 0]), atol=1e-6)
+    assert_equal(res.success, False)
+    assert_equal(res.status, 4)
+
+def test_bounded_powell_vs_powell():
+    # here we test an example where the bounded Powell method
+    # will return a different result than the standard Powell
+    # method.
+
+    # first we test a simple example where the minimum is at
+    # the origin and the minimum that is within the bounds is
+    # larger than the minimum at the origin.
+    func = lambda x: np.sum(x**2)
+    bounds = (-5, -1), (-10, -.1), (1, 9.2), (-4, 7.6), (-15.9, -2)
+    x0 = [-2.1, -5.2, 1.9, 0, -2]
+
+    options = dict(ftol=1e-10, xtol=1e-10)
+
+    res_powell = optimize.minimize(func, x0, method="Powell", options=options)
+    assert_allclose(np.array([0.]*len(x0)), res_powell.x, atol=1e-6)
+    assert_allclose(0., res_powell.fun, atol=1e-6)
+
+    res_bounded_powell = optimize.minimize(func, x0, options=options,
+                                           bounds=bounds,
+                                           method="Powell")
+    p = np.array([-1, -.1, 1, 0, -2])
+    assert_allclose(p, res_bounded_powell.x, atol=1e-6)
+    assert_allclose(func(p), res_bounded_powell.fun, atol=1e-6)
+
+    # now we test bounded Powell but with a mix of inf bounds.
+    bounds = (None, -1), (-np.inf, -.1), (1, np.inf), (-4, None), (-15.9, -2)
+    res_bounded_powell = optimize.minimize(func, x0, options=options,
+                                           bounds=bounds,
+                                           method="Powell")
+    p = np.array([-1, -.1, 1, 0, -2])
+    assert_allclose(p, res_bounded_powell.x, atol=1e-6)
+    assert_allclose(func(p), res_bounded_powell.fun, atol=1e-6)
```

Comment:
Needs tests for:

- all lower/upper bounds are -np.inf/np.inf
- where parameter bounds are one sided and the target parameter values are large (e.g. 1e6) (to test stability of `np.tan` transform), preferably with the problem having some parameters that have df_{i} / dx_{i} with large and some small, and that the correct solution is found. It needs to work across a range of problems. Hopefully this is not an onerous requirement, if you have a reference to other situations where this transform is employed it may not be as necessary.

Reply:
First one done. I will think a little about the second one. Also, see my response to the the other comment regarding arctan stability.

Reply:
I added the function `test_onesided_bounded_powell_stability` to `test_optimize.py`. I'm not totally sure if this is what you wanted, so please let me know. Again, please see my response to your previous comment about the `np.tan` stability.